extras/gizmo/gizmo.js

```diff
@@ -0,0 +1,330 @@
+        }
+        tmpV1.scale(1.0 / (this.nodes.length || 1));
+        this.gizmo.setPosition(tmpV1);
+
+        this.fire('position:set', tmpV1);
+    }
+
+    _updateRotation() {
+        tmpV1.set(0, 0, 0);
+        if (this._coordSpace === 'local') {
+            for (let i = 0; i < this.nodes.length; i++) {
+                const node = this.nodes[i];
+                tmpV1.add(node.getEulerAngles());
+            }
+            tmpV1.scale(1.0 / (this.nodes.length || 1));
+        }
+        this.gizmo.setEulerAngles(tmpV1);
+
+        this.fire('eulerAngles:set', tmpV1);
+    }
+
+    _updateScale() {
+        if (this.camera.projection === PROJECTION_PERSPECTIVE) {
+            this._scale = this._getProjFrustumWidth() * PERS_SCALE_RATIO;
+        } else {
+            this._scale = this.camera.orthoHeight * ORTHO_SCALE_RATIO;
+        }
+        this._scale = Math.max(this._scale * this._size, MIN_GIZMO_SCALE);
+        this.gizmo.setLocalScale(this._scale, this._scale, this._scale);
+
+        this.fire('scale:set', this._scale);
+    }
+
+    _getSelection(x, y) {
+        const start = this.camera.screenToWorld(x, y, 1);
+        const end = this.camera.screenToWorld(x, y, this.camera.farClip);
+        const dir = end.clone().sub(start).normalize();
+
+        const selection = [];
+        for (let j = 0; j < this.intersectMeshInstances.length; j++) {
+            const meshInstance = this.intersectMeshInstances[j];
+            const mesh = meshInstance.mesh;
+            const wtm = meshInstance.node.getWorldTransform().clone().invert();
+
+            wtm.transformPoint(start, xstart);
+            wtm.transformVector(dir, xdir);
+            xdir.normalize();
+
+            const pos = [];
+            const indices = [];
+            mesh.getPositions(pos);
+            mesh.getIndices(indices);
+
+            for (let k = 0; k < indices.length; k += 3) {
+                const i1 = indices[k];
+                const i2 = indices[k + 1];
+                const i3 = indices[k + 2];
+
+                tmpV1.set(pos[i1 * 3], pos[i1 * 3 + 1], pos[i1 * 3 + 2]);
+                tmpV2.set(pos[i2 * 3], pos[i2 * 3 + 1], pos[i2 * 3 + 2]);
+                tmpV3.set(pos[i3 * 3], pos[i3 * 3 + 1], pos[i3 * 3 + 2]);
+                tmpT1.set(tmpV1, tmpV2, tmpV3);
+
+                if (tmpT1.intersectRay(xstart, xdir)) {
+                    selection.push(meshInstance);
+                }
+            }
+
+        }
+        return selection;
+    }
+
+    /**
+     * Attach an array of graph nodes to the gizmo.
+     *
+     * @param {import('playcanvas').GraphNode} [nodes] - The graph nodes. Defaults to [].
+     * @example
+     * const gizmo = new pcx.Gizmo();
+     * gizmo.attach([boxA, boxB]);
+     */
+    attach(nodes = []) {
+        this.nodes = nodes;
+        this._updatePosition();
+        this._updateRotation();
+
+        window.addEventListener('pointerdown', this._onPointerDown);
```

Comment:
It's worth adding a [check to see if](https://github.com/playcanvas/engine/blob/4ce20f98907ac0395a1fbeaeda21acdfc7f7875f/src/core/platform.js#L84) `window` exists, just in case this somehow gets pulled into a non browser environment. Edge case I know.